Add subfield pruning for lambda for parametric functions (#21069)
prestodb/presto: presto-docs/src/main/sphinx/develop/functions.rst

            @BlockPosition @SqlType("T") Block block,
            @BlockIndex int index) { /* ...implementation */ }

Complex Type Function Descriptor
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Complex type function descriptor is used for specifying the metadata for the map or array functions that deal with arguments of ``ROW`` types. Presto optimizer tries to collect

Reviewer: ```suggestion
Complex type function descriptor is used for specifying the metadata for the map or array functions that deal with arguments of ``ROW`` types. The Presto optimizer tries to collect
```

Author: Applied

> information about accessed subfields as granular as possible so tha file format reader (ORC/DWRF/Parquet) could prune all subfields that are irrelevant for execution of the current

Reviewer: ```suggestion
information about accessed subfields as granular as possible so the file format reader (ORC/DWRF/Parquet) can prune all subfields that are irrelevant for execution of the current
```

Author: Applied

query. Consider the example below.

> .. code-block:: sql

Reviewer: Please fix the code block, and if you can, build the docs locally and review them for correct formatting. 
(See "Building manually" and "Viewing the documentation" in the [presto-docs README](https://github.com/prestodb/presto/tree/master/presto-docs) for how to build and view the docs locally. )

A local build of this doc looks like this: 
<img width="958" alt="Screenshot 2024-01-02 at 2 03 17 PM" src="https://github.com/prestodb/presto/assets/7013443/f23e9593-5b38-4366-b14d-4c135fcfed09">

Author: fixed

CREATE TABLE my_table (array_of_structs array(row(a bigint, b varchar, c double, d row(d1 bigint, d2 double))))

SELECT CARDINALITY(FILTER(array_of_structs, x -> x.a > 0)) FROM my_table

SELECT FILTER(array_of_structs, x -> x.a > 0) FROM my_table

> In query ``SELECT CARDINALITY(FILTER(array_of_structs, x -> x.a > 0)) FROM my_table``, we need only ``array_of_structs[*].x`` subfield for executing the query. However, in query

Reviewer: ```suggestion
In the query ``SELECT CARDINALITY(FILTER(array_of_structs, x -> x.a > 0)) FROM my_table``, we need only the ``array_of_structs[*].x`` subfield to execute the query. However, in the query
```

Author: Applied

> ``SELECT FILTER(array_of_structs, x -> x.a > 0) FROM my_table`` we need all subfields of column y. That is due to the difference in intrinsic properties of these functions.

Reviewer: ```suggestion
``SELECT FILTER(array_of_structs, x -> x.a > 0) FROM my_table`` we need all of the subfields of column ``y``,  due to the difference in intrinsic properties of these functions.
```

Author: Applied

> Complex type function descriptor captures those properties.

Reviewer: ```suggestion
The complex type function descriptor captures those properties.
```

Author: Applied

Complex type function descriptor could be defined for ``@ScalarFunction`` and ``@CodegenScalarFunction`` annotation using ``descriptor`` parameter.

> .. code-block:: java

Reviewer: Please fix this code block.

Author: fixed

@Description("Returns true if the array contains one or more elements that match the given predicate")
@ScalarFunction(value = "any_match", descriptor = @ScalarFunctionDescriptor(
        isAccessingInputValues = true,
        argumentIndicesContainingMapOrArray = {},
        outputToInputTransformationFunction = "clearRequiredSubfields",
        lambdaDescriptors = {
                    @ScalarFunctionLambdaDescriptor(
                        lambdaArgumentDescriptors = {
                                @ScalarFunctionLambdaArgumentDescriptor(callArgumentIndex = 0)})}))
public final class ArrayAnyMatchFunction
{
    ...
}

* ``isAccessingInputValues``:

>   Indicates whether the function accesses subfields. If function accesses the subfields internally, then optimizer will check ``lambdaDescriptors`` in order to collect all

Reviewer: ```suggestion
  Indicates whether the function accesses subfields. If the function accesses the subfields internally, then the Presto optimizer checks ``lambdaDescriptors`` to collect all
```

Author: Applied

>   accessed subfields. If ``lambdaDescriptors``is empty, then it disqualify the function from the optimization. In this case, optimizer will assume that all subfields are required.

Reviewer: ```suggestion
  accessed subfields. If ``lambdaDescriptors``is empty, then it disqualifies the function from the optimization. In this case, the Presto optimizer assumes that all subfields are required.
```

Author: Applied

  Default value is ``true``.

* ``lambdaDescriptors``:
  Contains the array of ``@ScalarFunctionLambdaDescriptor`` for each lambda that this function accepts.
  ``@ScalarFunctionLambdaDescriptor`` contains only one property:

  * ``lambdaArgumentDescriptors``:
    Contains the array of ``@ScalarFunctionLambdaArgumentDescriptor`` for each argument that lambda accepts.
    ``@ScalarFunctionLambdaArgumentDescriptor`` contains two properties:

    * ``@callArgumentIndex``:
      Index of the function argument that contains the function input (Array or Map), to which this lambda argument relate to.

>       Example 1: ``SELECT ANY_MATCH(array_of_structs, x -> x.a > 0) FROM my_table``. The lambda passed to the ``ANY_MATCH`` function will be invoked internally in ``ANY_MATCH``

Reviewer: ```suggestion
      Example 1: ``SELECT ANY_MATCH(array_of_structs, x -> x.a > 0) FROM my_table``. The lambda passed to the ``ANY_MATCH`` function is invoked internally in ``ANY_MATCH``
```

Author: Applied

      function passing the element of the array, which ``ANY_MATCH`` receives in the first argument (with index 0 in a 0-indexed array of
      function arguments) of the ``ANY_MATCH`` function.
      Therefore, ``@callArgumentIndex`` of the ``@ScalarFunctionLambdaArgumentDescriptor`` for the  ``ANY_MATCH`` function should be `0`.

>       Example 2: ``SELECT MAP_FILTER(map_struct_to_struct, (key, value) -> key.field_k > 0 && value.field_v > 0) FROM my_table``. The lambda passed to the ``ANY_MATCH`` function will be invoked internally in ``ANY_MATCH`` function

Reviewer: ```suggestion
      Example 2: ``SELECT MAP_FILTER(map_struct_to_struct, (key, value) -> key.field_k > 0 && value.field_v > 0) FROM my_table``. The lambda passed to the ``ANY_MATCH`` function is invoked internally in ``ANY_MATCH`` function
```

Author: Applied

      passing the element of the array, which ``ANY_MATCH`` receives in the first argument (with index 0 in a 0-indexed array of
      function arguments) of the ``ANY_MATCH`` function.
      Therefore, ``@callArgumentIndex`` of the ``@ScalarFunctionLambdaArgumentDescriptor`` for the  ``ANY_MATCH`` function should be `0`.

    * ``@lambdaArgumentToInputTransformationFunction``:
      Contains the transformation function between the subfields of this lambda argument and the input of the function.

>       Default value for this parameter is ``"prependAllSubscripts"``, will add prefix ``[*]`` to the

Reviewer: ```suggestion
      Default value for this parameter is ``"prependAllSubscripts"``, which adds prefix ``[*]`` to the
```

Author: Applied

      path of the subfield, which is correct for 99% of the functions.
      Example: ``SELECT ANY_MATCH(array_of_structs, x -> x.a > 0) FROM my_table``. Even though in the lambda the access subfield has path `a`,
      we need to add the prefix ``[*]``, because ``array_of_structs`` is of array type.

  Default value of the ``lambdaDescriptors`` is an empty array. If function does not accept any lambda parameter, then ``lambdaDescriptors`` should be an empty array.

* ``argumentIndicesContainingMapOrArray``:
  Consider ``TRIM_ARRAY`` function in this example: ``SELECT ANY_MATCH(TRIM_ARRAY(array_of_structs, 5), x -> x.a > 0) FROM my_table``.

>   By default, the optimizer passes the subfield accessed in outer functions (here, subfield ``[*].a``, which accessed in function ``ANY_MATCH``) to all array or map arguments

Reviewer: ```suggestion
  By default, the optimizer passes the subfield accessed in outer functions (here, subfield ``[*].a``, which is accessed in function ``ANY_MATCH``) to all array or map arguments
```
Is this suggestion correct?

Author: Applied

  (here, to the array ``array_of_structs``).

>   Now, consider function ``MAP`` in this example below.

Reviewer: ```suggestion
  Consider the function ``MAP`` in this example:
```

Author: Applied

>   .. code-block:: sql

Reviewer: Please fix this code block.

Author: fixed

  CREATE TABLE my_table (keys array(bigint), values array(row(a bigint, b varchar, c double, d row(d1 bigint, d2 double))))

  SELECT ANY_MATCH(MAP_VALUES(MAP(keys, values)), x -> x.a > 100)

  In ``MAP`` function, both ``keys`` and ``values`` are of array type. However, subfields accessed in function ``ANY_MATCH`` relate only to ``values``.  Note, for map type,

>   we collect only subfields of map values.  Default behavior does not work here and we need to manually specify for ``MAP`` function that subfields accessed in outer functions

Reviewer: ```suggestion
  we collect only subfields of map values.  Default behavior does not work here and we must manually specify for ``MAP`` function that subfields accessed in outer functions
```

Author: Applied

  relate only to the second argument (with index 1 in 0-indexed array of arguments).
  Therefore, for ``MAP`` function we have to specify``argumentIndicesContainingMapOrArray = {1}``.

>   Unless the function does not have such behaviour like ``MAP`` function, ``argumentIndicesContainingMapOrArray`` parameter could be omitted.

Reviewer: Could "Unless" be "If" here?

Author: Changed to "If"

* ``outputToInputTransformationFunction``:
  Contains the transformation function to convert the output back to the input elements of the array or map. Optimizer examines the nested function calls from the most outer

>   function to the most inner function. Oftentimes, the path of the subfields need to be altered reverting the effect that is done by the function to its input. Let's consider the

Reviewer: ```suggestion
  function to the most inner function. Often the path of the subfields must be altered, reverting the effect that is done by the function to its input. Let's consider the
```

Author: Applied

  example:

  .. code-block:: sql
  CREATE TABLE my_table (2d_array_of_structs array(array(row(p bigint, e row(e1 bigint, e2 varchar)))))

  SELECT TRANSFORM(FLATTEN(2d_array_of_structs), x -> x.p) FROM my_table

  Here, column ``2d_array_of_structs`` is a 2-dimensional array or rows. Function ``FLATTEN`` produces the 1-dimensional array of rows. In ``TRANSFORM``,
  subfield ``[*].p`` is accessed. The challenge now is to correctly attribute accessed subfield ``[*].p`` to the original column ``2d_array_of_structs`` where the path of the
  subfield `p` is ``[*][*].p``. This is when ``outputToInputTransformationFunction`` comes into play. Because of the internal property of the ``FLATTEN`` function,

>   for any subfield that accessed in outer calls we need to transform the path of the subfield and add additional ``[*]`` prefix.

Reviewer: ```suggestion
  for any subfield that accessed in outer calls we must transform the path of the subfield and add additional ``[*]`` prefix.
```

Author: Applied

  There are several pre-defined values for ``outputToInputTransformationFunction`` parameter:

  * ``"identity"``:
    Identity function (no transformation needed).

  * ``"allSubfieldsRequired"``:

>     This transformation function will indicated that all subfield are required. At this point, optimizer will discard any collected subfields from outer calls and add special

Reviewer: ```suggestion
    This transformation function indicates that all subfields are required. At this point, the Presto optimizer discards any collected subfields from outer calls and adds the special
```

Author: Applied

    subfield ``[*]``  (allSubfields). It means that the transformation of the output to input is unknown and thus the lambda subfields pushdown from outer calls could not be
    done. This is a default value of ``@ScalarFunction`` annotation.

  * ``"clearRequiredSubfields"``:

>     There are some functions that do not send the input to the output. For instance, ``CARDINALITY`` function, does not returns the input array like ``FILTER`` function does.

Reviewer: ```suggestion
    There are some functions that do not send the input to the output. For instance, ``CARDINALITY`` function, does not return the input array like the ``FILTER`` function does.
```

Author: Applied

    Instead, it returns only the number (the size of the array). Knowing this property of the function, we can safely discard any accessed subfields in outer functions (even

>     ``[*]``) that essentially means that optimizer can start collecting the accessed subfields from now on. If we are lucky and any other inner call does not include

Reviewer: ```suggestion
    ``[*]``) that essentially means that the Presto optimizer can start collecting the accessed subfields from now on. If we are lucky and any other inner call does not include
```

Author: Applied
